File: 백준/Silver/2992. 크면서 작은 수/크면서 작은 수.py
# 크면서 작은 수
# 같은 글자 조합을 가지고 있으면서, 현재값 보단 큰 수 중에서 가장 작은 값을 구하는 문제
# 이건 조합 구현하는 방법을 한번 찾아보는게 좋을듯
# 4글자라고 하면 4C4를 전부 구하고 거기서 반복문을 돌려 기존값 보다 큰 바로 다음 값을 구하면 되는 문제
# 조합 공식이 뭐징

# 정수 X 받기((리스트로 받을게유)
X = input()

# 정수 X로 만들수 있는 조합 담는 배열
combination = []

# X의 문자 조합들 구하기
# 뭔가 잘못된거 같아 조합의 수가 너무 많이 나와
# 27711이 중복된 숫자들이 있어서 그런진 모르겠는데 중복된 결과들이 너무 많이 나오는데 이게 제대로 된 조합들인지, 아니면 조합이 중복으로 생성되는건지 모르겠네
def get_comb(X, now_idx, word, length, visited):
    # 이미 사용된 인덱스면 중단
    if not visited[now_idx]:
        return

    # 사용여부 체크
    visited[now_idx] = False

    # 문자 추가
    word += X[now_idx]
    
    # 길이가 X만큼인 문자열이 완성되었으면 combination에 추가
    if length == len(X):
        combination.append(word)
        return
    
    # 지금 문자 빼고 문자 싹 다 집어넣기
    for next_idx in range(len(X)):
        if next_idx != now_idx:
            # visited 참조를 다르게 사용해야 하니 스프레드 연산자 써서 얕은복사
            get_comb(X, next_idx, word, length + 1, [*visited])

# combination에서 결과값 찾는 함수
def get_result(X, combination):
    # combination에서 X보다 큰 수 나오면 그거 출력하고 종료
    for now in combination:
        if int(X) < int(now):
            return now
    return 0

# combination 구하기
for i in range(len(X)):
    # 사용한 인덱스 체크해주는 배열
    visited = [True] * len(X)
    # get_comb 실행
    get_comb(X, i, '', 1, visited)

# 에라 모르겠다 중복 겁나 많은거 set으로 없애주고 정렬해서 구하면 답은 나옴
combination = [*set(combination)]
combination.sort()

# 나중에 시간복잡도도 고려한 코드 보고 공부해보자

# 결과 출력
print(get_result(X, combination))

File: 백준/Silver/2992. 크면서 작은 수/크면서 작은 수.java
// 크면서 작은 수
// 같은 글자 조합을 가지고 있으면서, 현재값 보단 큰 수 중에서 가장 작은 값을 구하는 문제
// 이건 조합 구현하는 방법을 한번 찾아보는게 좋을듯
// 4글자라고 하면 4C4를 전부 구하고 거기서 반복문을 돌려 기존값 보다 큰 바로 다음 값을 구하면 되는 문제
// 조합 공식이 뭐징

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;

public class Main {
    // 정수 X로 만들수 있는 조합 담는 배열
    // 이렇게 class 바로 상위에 static으로 선언하면 전역변수로 사용 가능함
    static ArrayList<String> combinations = new ArrayList<>();

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String X = br.readLine();

        for(int i = 0; i < X.length(); i++){
            // 사용 여부 체크
            boolean[] visited = new boolean[X.length()];
            Arrays.fill(visited, true);

            // 조합 구하는 DFS 실행
            getComb(X, i, "", 1, visited);
        }

        System.out.println(getResult(X, combinations));
    }

    // combination에서 결과값 찾는 함수
    public static void getComb(String X, int now_idx, String word, int length, boolean[] visited){
        // 사용여부 확인
        if(!visited[now_idx]){
            return;
        }

        // 사용 체크
        visited[now_idx] = false;

        // 현재 문자 추가
        word += X.charAt(now_idx);

        // 문자열 완성됐으면 combination에 추가
        if(length == X.length()){
            combinations.add(word);
            return;
        }

        for(int next_idx = 0; next_idx < X.length(); next_idx++){
            if(next_idx != now_idx){
                // 자바 얕은복사는 clone()이란 메서드가 있네
                // 뭔 파이썬, JS, 자바 전부 다 달라 미쳐버리겠네;;
                getComb(X, next_idx, word, length + 1, visited.clone());
            }
        }

    }

    // 결과값 찾는 함수
    public static int getResult(String X, ArrayList<String> combinations){
        // 오름차순 정렬
        combinations.sort(null);
        // 정수 X
        int int_X = Integer.parseInt(X);
        for(int i = 0; i < combinations.size(); i++){
            // 정수 조합
            int int_combination = Integer.parseInt(combinations.get(i));

            // 둘이 비교해서 X보다 큰 가장 첫번째 수를 반환하고 종료
            if(int_X < int_combination){
                return int_combination;
            }
        }
        return 0;
    }
}
